### PR TITLE
Improve performance of external (UDR) functions

### PR DESCRIPTION
**Pull request by asfernandes**
Test preparation:
```
recreate function div_udr (
    n1 integer,
    n2 integer
) returns double precision
    external name 'udf_compat!UC_div'
    engine udr;

create domain d_int_check integer check (value > 0);
create domain d_double_check double precision check (value >= 0);

recreate function div_udr2 (
    n1 d_int_check,
    n2 d_int_check
) returns d_double_check
    external name 'udf_compat!UC_div'
    engine udr;
```

Comparison:
```
set stats on;

-- 4.4s -> 2.7s
execute block
as
    declare n integer = 5000000;
    declare x integer;
begin
    while (n > 0) do
    begin
        n = n - 1;
        x = div_udr(3, 2);
    end
end;

-- 5.1s -> 3.2s
execute block
as
    declare n integer = 5000000;
    declare x integer;
begin
    while (n > 0) do
    begin
        n = n - 1;
        x = div_udr2(3, 2);
    end
end;
```

### CODE REVIEW

**Comment by sim1984** [comment]
Does this only improve the performance of external functions only, or is there an effect for procedures/triggers too?

**Comment by asfernandes** [comment]
> Does this only improve the performance of external functions only, or is there an effect for procedures/triggers too?

It's mostly about functions only.

There could be a marginal improvement with procedures with parameters with constraints, but I didn't tested it, as procedure/trigger implementation could also be greater improved, but is much more difficult.

**Comment by asfernandes** [comment]
@dyemanov are you going to review this before merge?

File: src/jrd/ExtEngineManager.cpp [hunk 1]
```diff
@@ -774,20 +901,124 @@ ExtEngineManager::Function::~Function()
...
+	{
+		const auto extInMsg = request->getImpure<UCHAR>(extInputImpureOffset.value());
+		copyMessage(tdbb, udf->getInputFormat(), inMsg, extInputFormat, extInMsg);
+		inMsg = extInMsg;
+		///inMsgLength = extInputFormat->fmt_length;
+	}
+
+	// Initialize outputs in the internal message.
+	{	// scope
+		fb_assert(udf->getOutputFormat()->fmt_desc.getCount() / 2 == udf->getOutputFields().getCount());
```
Comment (dyemanov) [inline src/jrd/ExtEngineManager.cpp:930]: I see nothing this scope would protect. Is it really needed?

File: src/jrd/ExtEngineManager.cpp [hunk 2]
```diff
@@ -774,20 +901,124 @@ ExtEngineManager::Function::~Function()
...
+
+			dsc* defaultValue = nullptr;
+
+			if (request->getStatement()->mapFieldInfo.get(namePair, fieldInfo) &&
+				fieldInfo.defaultValue)
+			{
+				defaultValue = EVL_expr(tdbb, request, fieldInfo.defaultValue);
+
+				if (request->req_flags & req_null)
+					defaultValue = nullptr;
```
Comment (dyemanov) [inline src/jrd/ExtEngineManager.cpp:949]: IIRC, this condition/assignment is redundant, `EVL_expr()` is guaranteed to return `nullptr` in this case.

File: src/jrd/ExtEngineManager.cpp [hunk 3]
```diff
@@ -65,21 +67,74 @@ static EngineCheckout::Type checkoutType(IExternalEngine* engine);
...
+
+			if (!*srcNullPtr)
+			{
+				dsc srcDesc = srcDescIt[0];
+				srcDesc.dsc_address = srcMsg + srcArgOffset;
+
+				dsc dstDesc = dstDescIt[0];
+				dstDesc.dsc_address = dstMsg + dstArgOffset;
+
+				MOV_move(tdbb, &srcDesc, &dstDesc);
```
[the remark this answers is not in the document]
Comment (asfernandes) [reply src/jrd/ExtEngineManager.cpp:111]: It's already being done in CVT_move_common.
It's not inline, but, I think this is not going to be so relevant.

File: src/jrd/ExtEngineManager.cpp [hunk 4]
```diff
@@ -65,21 +67,74 @@ static EngineCheckout::Type checkoutType(IExternalEngine* engine);
...
+
+				*dstNullPtr = 0;
+			}
+			else
+				*dstNullPtr = -1;
```
[the remark this answers is not in the document]
Comment (asfernandes) [reply src/jrd/ExtEngineManager.cpp:116]: I think `FB_FALSE` / `FB_TRUE` could be used as we used them extensively in similar places.

File: src/jrd/ExtEngineManager.cpp [hunk 5]
```diff
@@ -150,99 +205,103 @@ namespace
...
-			  argNumber(aArgNumber)
+		InitParametersNode(thread_db* tdbb, MemoryPool& pool, CompilerScratch* csb,
+				Array<NestConst<Parameter>>& parameters, MessageNode* aMessage)
+			: TypedNode<StmtNode, StmtNode::TYPE_EXT_INIT_PARAMETERS>(pool),
+			  message(aMessage)
 		{
-			Parameter* parameter = parameters[argNumber / 2];
-			defaultValueNode = NULL;
+			// Iterate over the format items, except the EOF item.
+			const unsigned paramCount = message->format->fmt_count / 2;
```
[the remark this answers is not in the document]
Comment (asfernandes) [reply src/jrd/ExtEngineManager.cpp:222]: It should be wrong. It's used to initialize output message, that is going to have the null flag.
In this case, it purposely ignores it.

File: src/jrd/ExtEngineManager.cpp (hunk 1 above)
Comment (asfernandes) [reply src/jrd/ExtEngineManager.cpp:930]: I used it mostly as a block to not create a separate function just for it.
I'm not removing the explicit `scope` label.

**Review by dyemanov — APPROVED** [review]

**Pull request merged**